### ISSUE TITLE
Create more Sample Projects

### ISSUE THREAD

**Issue by AdamsLair** [AdamsLair/duality#211]
As described [here](http://forum.adamslair.net/viewtopic.php?f=4&t=555), it would be a good idea to thoroughly test v2.0 by implementing some more sample projects.

---------

**Note:** If you're contributing to this issue, the default should be to create your own, distinct sample repository, rather than working on the official one and creating Pull Requests. You can still release your sample packages using NuGet and they will show up in the Duality Package Manager / Sample Browser.

NuGet Docs: [here](https://docs.nuget.org/create)
Package Specs (Look at the `.Sample` ones): [here](https://github.com/AdamsLair/duality/tree/master/NuGetPackageSpecs)

**Note:** Sample Projects with more than 1 MB content size in total (Data directory) shouldn't go into the main repo at all, but into a distinct sample repo. As a general rule of thumb, a main repo sample ideally should stay below 500 KB to keep the history small. None of this applies for sample projects in distinct repositories - which is another reason why they should be preferred.

**Comment by AdamsLair** [comment]
### Single-Topic Ideas

- Input Management
- Audio Management
- Basic Shaders
- Multi-Pass Rendering and Postprocessing
- Scene Transitions
- A Basic Particle System

### Multi-Topic Ideas

- Basic Menu Screen
- Physics-Based Platformer
- RTS Game Basics

### Example Game ideas

- "Flappy Bird" Clone
- Minimalistic "Angry Birds" Clone

**Comment by HeadClot** [comment]
A Top Down Example

**Comment by RockyTV** [comment]
Don't know if this would be a good suggestion, but a mandelbrot fractal generator could be neat. It could also serve as a performance test.

**Comment by ilexp** [comment]
**Progress:**

- Added an Input Handling sample package that also acts as an input testbed.
- Fixed some input bugs in the process and also added support for OpenTK's new "Joystick Hat" detection.

**Comment by hsnabn** [comment]
How about a sample to demonstrate how to do timed actions (triggers) in Duality?
Here is a thread on the AdamsLair Forums where SirePi and I had a similar discussion:
http://forum.adamslair.net/viewtopic.php?f=4&t=515

**Comment by hsnabn** [comment]
I found a link to a basic C# particle system:
http://www.codeproject.com/Articles/10003/A-basic-Particles-System
It does not deal with any graphics, it's only a sort of backend.
It would take some work to port it over, I think. I don't know if I am good enough at programming to port it over to Duality.
Of course, there may be other ways to implement a particle system. I'm looking at you, @SirePi.

**Comment by ilexp** [comment]
> I found a link to a basic C# particle system:

The Dual Stick Space Shooter example already includes a custom particle system - I might just port it over into a distinct sample at some point :)

**Comment by SirePi** [comment]
@hsnabn: yeah, I should probably make some samples for all my plugins :smile: just let me finish the porting to Duality 2.0 and you'll see
@ilexp: I've begun to make the Flappy Bird clone.. I call it "Flap Or Die" :laughing:

**Comment by ilexp** [comment]
@SirePi I've looked into your Basic Shaders sample and would propose some changes - however, it's probably faster if I'll just do it myself in this case (also, given [this](https://github.com/AdamsLair/duality/issues/218) which I'll probably do in the process). Is it fine for you when I pull your branch locally, apply my (destructive, in this case) changes and push it back to master? If not, I'd just quickly set up my own shader sample. 

> @ilexp: I've begun to make the Flappy Bird clone.. I call it "Flap Or Die" :laughing: 

:smile: 

Added you to the list.

**Comment by SirePi** [comment]
@ilexp I sent the PR. Feel free to alter it, tear it apart or do whatever you feel is better

**Comment by ilexp** [comment]
> @ilexp I sent the PR. 

I've closed the PR again, because I planned to pull this manually using my local git, rather than merging first and adjusting later.

> Feel free to alter it, tear it apart or do whatever you feel is better

Thanks!

**Comment by SirePi** [comment]
Ah ok. tell me when it's done so I can delete the branch

**Comment by ilexp** [comment]
> Ah ok. tell me when it's done so I can delete the branch

Done! Merged your branch and added some changes on top of it. Feel free to delete your local branch. :)

**Comment by ilexp** [comment]
**Progress:**

- Added an Audio Handling sample package.
- Fixed some audio bugs in the process and also added support for lowpass filtering to allow users to dynamically simulate sounds heard from "behind a wall", like the rain outside, music from inside a pub or similar.

**Comment by SirePi** [comment]
Ooh, leave the particles for me.. I can simply extract the code from my plugin and just add some minor changes :smile:

**Comment by ilexp** [comment]
> Ooh, leave the particles for me.. I can simply extract the code from my plugin and just add some minor changes :smile:

I kind of wanted that one, too :smile: Compromise: You do it, and we'll decide in the review whether it will be an inofficial one (Your repo, you release it yourself) or an official one (Merge, I'll take it from there) - if that would be fine for you?

> Also the Game of life is interesting.. what do you mean by "shaders driven"?

I thought about doing ping-pong rendering between two RenderTargets and a pixel shader which does all the Game of Life calculations. That way, the game could be ridiculously large without being slow.

**Comment by SirePi** [comment]
> I kind of wanted that one, too
Oh, no worries, have fun with that :smiley: 
> I thought about doing ping-pong rendering between two RenderTargets and a pixel shader which does all the Game of Life calculations. That way, the game could be ridiculously large without being slow.
Yours as well then :laughing: 

I could do the Timed Events one then?

**Comment by hsnabn** [comment]
I actually started on that already, but go ahead; your sample will probably be better than mine, and I haven't actually gone very far.

**Comment by SirePi** [comment]
> I actually started on that already, but go ahead; your sample will probably be better than mine, and I haven't actually gone very far.

Gah! is there anything left for me then? :cry:

**Comment by hsnabn** [comment]
Go ahead and start working on the Timed Actions sample. Like I said, I haven't gotten far, and I don't really code in quality.

**Comment by SirePi** [comment]
Well, the Flappy Bird clone is coming along.. the game is basically done by itself, now it needs a good coat of graphics and music.

Check it [here](https://github.com/SirePi/duality/tree/flap-or-die-sample)

**Comment by ilexp** [comment]
> I think I'll implement a high score mechanism based on UserData as well...

UserData isn't really intended for that - can you just use `DualityApp.System.GetNamedPath(...)` to get the user directory and put your highscore in there using `Serializer.WriteObject(...)`? That way, you'd be showing off how to write chunks of arbitrary data somewhere - like for a highscore list :)

**Comment by hsnabn** [comment]
A little off-topic here, but where would be the best location to store game save data? I don't really like storing data in the user directory, so I store the data within the application folder (the `DualityApp.Data` directory). But, I think this might cause problems with some users (Windows UAC), so is the user directory the most advisable location?

**Comment by ilexp** [comment]
> A little off-topic here, but where would be the best location to store game save data? I don't really like storing data in the user directory, so I store the data within the application folder (the `DualityApp.Data` directory). But, I think this might cause problems with some users (Windows UAC), so is the user directory the most advisable location?

I'm not sure UAC will block control if an application tries to write a new file inside its own folder structure, especially if it's not an installed program and thus will likely not be located in any of the regular Windows folders (Programs, etc.) - never had a problem with this, though I'm not really an expert on UAC permissions.

**Comment by hsnabn** [comment]
Yeah - otherwise, a lot of applications wouldn't be able to write logs and such, anyways.
Thanks for the reply.

**Comment by hsnabn** [comment]
I built a Gravity sample for Dualtiy 2.0
It's located in my samples repository, [here](github.com/hsnabn/duality-samples).
It shows how to make objects exert forces on other objects.
The sample includes a small test scene where you can add objects and do some limited tweaking of some values.

**Comment by hsnabn** [comment]
I'm currently working on a RigidBody-based Platformer sample.

I've managed to work in some nice Finite State Machine -ish logic into it. It is based on a State `enum` and a `switch` statement. I got lots of help from here: http://gameprogrammingpatterns.com/state.html

Of course, it's still rudimentary. It still relies on conditional logic a lot. I am not very good at it, so it was inevitable.

I've just finished creating the player's movement logic and basic state management. I don't know if I will be including any "enemy" characters in this one, though.

I'll be pushing it to my samples repository after I get in at least one level. Maybe I'll add levels, maybe not. I'm actually going to use this as a base for my next project, which is coincidentally also a platformer.
